File: results/dnd-rest-benchmark/lifecycle-runs/20260804T072004Z_claude_claude-sonnet-5_ruby-sinatra/routes/play/invitations.rb
# Campaign invitations: a DM invites a registered player identity to join
# the campaign with a specific character_id, and only that identity may
# accept. Acceptance mirrors the regular join flow (routes/play/campaigns.rb
# POST /members) so an accepted invitation leaves the campaign in the same
# state a self-service join would have.

def play_invitation_payload(row)
  {
    invitation_id: row['invitation_id'],
    username: row['username'],
    character_id: row['character_id'],
    status: row['status']
  }
end

def next_play_invitation_sequence(campaign_id)
  db.execute(
    'SELECT COALESCE(MAX(sequence), 0) + 1 AS n FROM play_invitations WHERE campaign_id = ?',
    [campaign_id]
  ).first['n']
end

def find_play_invitation!(campaign_id, invitation_id)
  invitation = db.execute(
    'SELECT * FROM play_invitations WHERE campaign_id = ? AND invitation_id = ?',
    [campaign_id, invitation_id]
  ).first
  halt 404, { error: 'invitation not found' }.to_json unless invitation
  invitation
end

post '/v1/play/campaigns/:id/invitations' do
  user = authenticate_play_request!
  campaign = find_play_campaign!(params[:id])
  require_play_owner!(campaign, user)

  body = json_body
  halt 400, { error: 'invalid json' }.to_json if body.nil?

  invitation_id = body['invitation_id']
  username = body['username']
  character_id = body['character_id']

  halt 400, { error: 'invalid invitation_id' }.to_json unless invitation_id.is_a?(String) && !invitation_id.empty?
  halt 400, { error: 'invalid username' }.to_json unless username.is_a?(String) && !username.empty?
  halt 400, { error: 'invalid character_id' }.to_json unless character_id.is_a?(String) && !character_id.empty?

  target = db.execute('SELECT * FROM users WHERE username = ?', [username]).first
  halt 400, { error: 'invalid username' }.to_json unless target && target['role'] == 'player'

  existing = db.execute(
    'SELECT 1 FROM play_invitations WHERE campaign_id = ? AND invitation_id = ?',
    [campaign['id'], invitation_id]
  ).first
  halt 409, { error: 'invitation_id already exists' }.to_json if existing

  active = db.execute(
    "SELECT 1 FROM play_invitations WHERE campaign_id = ? AND username = ? AND status = 'pending'",
    [campaign['id'], username]
  ).first
  halt 409, { error: 'invitation already pending for this user' }.to_json if active

  sequence = next_play_invitation_sequence(campaign['id'])
  db.execute(
    'INSERT INTO play_invitations (campaign_id, sequence, invitation_id, username, character_id, status) ' \
    "VALUES (?, ?, ?, ?, ?, 'pending')",
    [campaign['id'], sequence, invitation_id, username, character_id]
  )

  status 201
  { invitation_id: invitation_id, username: username, character_id: character_id, status: 'pending' }.to_json
end

post '/v1/play/campaigns/:id/invitations/:invitation_id/accept' do
  user = authenticate_play_request!
  campaign = find_play_campaign!(params[:id])
  invitation = find_play_invitation!(campaign['id'], params[:invitation_id])

  halt 403, { error: 'forbidden' }.to_json unless invitation['username'] == user['username']
  halt 409, { error: 'invitation already accepted' }.to_json unless invitation['status'] == 'pending'

  character_id = invitation['character_id']

  db.execute(
    'INSERT INTO play_campaign_members (campaign_id, username, character_id, name, class) VALUES (?, ?, ?, ?, ?)',
    [campaign['id'], user['username'], character_id, user['username'], 'Adventurer']
  )
  db.execute(
    'INSERT INTO play_character_owners (campaign_id, character_id, owner) VALUES (?, ?, ?)',
    [campaign['id'], character_id, user['username']]
  )
  db.execute(
    'INSERT INTO play_character_gold (campaign_id, character_id, gold) VALUES (?, ?, ?)',
    [campaign['id'], character_id, 10]
  )
  db.execute(
    "UPDATE play_invitations SET status = 'accepted' WHERE campaign_id = ? AND invitation_id = ?",
    [campaign['id'], invitation['invitation_id']]
  )

  { invitation_id: invitation['invitation_id'], username: invitation['username'], character_id: character_id, status: 'accepted' }.to_json
end

get '/v1/play/campaigns/:id/invitations' do
  user = authenticate_play_request!
  campaign = find_play_campaign!(params[:id])

  is_owner = campaign['owner'] == user['username']

  rows = db.execute(
    'SELECT * FROM play_invitations WHERE campaign_id = ? ORDER BY sequence ASC',
    [campaign['id']]
  )

  records = if is_owner
              rows
            else
              rows.select { |row| row['username'] == user['username'] }
            end

  { invitations: records.map { |row| play_invitation_payload(row) } }.to_json
end

File: results/dnd-rest-benchmark/lifecycle-runs/20260731T160058Z_claude_claude-sonnet-5_typescript-vite/src/domain/play/invitations.ts
/**
 * Campaign invitations: a DM invites a registered player identity to join
 * with a specific character_id, and only that player may accept.
 */

import { getDb } from '../../db.ts';
import type { ApiResult, JsonValue } from '../../types.ts';
import { getUserRole } from '../auth.ts';
import {
  authenticate,
  isActor,
  isApiResult,
  findCampaign,
  requireParticipant,
  requireNonEmptyString,
} from './shared.ts';

interface InvitationRow {
  sequence: number;
  invitation_id: string;
  username: string;
  character_id: string;
  status: string;
}

interface InvitationObject {
  invitation_id: string;
  username: string;
  character_id: string;
  status: string;
}

function toInvitationObject(row: InvitationRow): InvitationObject {
  return {
    invitation_id: row.invitation_id,
    username: row.username,
    character_id: row.character_id,
    status: row.status,
  };
}

export function createInvitation(authHeader: string | undefined, campaignId: string, body: JsonValue): ApiResult {
  const actor = authenticate(authHeader);
  if (!isActor(actor)) return actor;

  const db = getDb();
  const campaign = findCampaign(db, campaignId);
  if (isApiResult(campaign)) return campaign;

  if (actor.username !== campaign.owner) {
    return { status: 403, body: { error: 'only the campaign dm may create invitations' } };
  }

  const invitationId = requireNonEmptyString(body.invitation_id, 'invitation_id');
  if (isApiResult(invitationId)) return invitationId;

  const username = requireNonEmptyString(body.username, 'username');
  if (isApiResult(username)) return username;

  const characterId = requireNonEmptyString(body.character_id, 'character_id');
  if (isApiResult(characterId)) return characterId;

  if (getUserRole(username) !== 'player') {
    return { status: 400, body: { error: 'username must be a registered player' } };
  }

  const existing = db
    .prepare('SELECT invitation_id FROM play_campaign_invitations WHERE campaign_id = ? AND invitation_id = ?')
    .get(campaignId, invitationId);
  if (existing) {
    return { status: 409, body: { error: 'invitation_id already exists' } };
  }

  const activeForUser = db
    .prepare(
      "SELECT invitation_id FROM play_campaign_invitations WHERE campaign_id = ? AND username = ? AND status = 'pending'",
    )
    .get(campaignId, username);
  if (activeForUser) {
    return { status: 409, body: { error: 'an active invitation already exists for this user' } };
  }

  const sequenceRow = db
    .prepare('SELECT COALESCE(MAX(sequence), 0) AS max_sequence FROM play_campaign_invitations WHERE campaign_id = ?')
    .get(campaignId) as { max_sequence: number };

  db.prepare(
    'INSERT INTO play_campaign_invitations (campaign_id, sequence, invitation_id, username, character_id, status) VALUES (?, ?, ?, ?, ?, ?)',
  ).run(campaignId, sequenceRow.max_sequence + 1, invitationId, username, characterId, 'pending');

  return {
    status: 201,
    body: {
      invitation_id: invitationId,
      username,
      character_id: characterId,
      status: 'pending',
    } as unknown as JsonValue,
  };
}

export function acceptInvitation(
  authHeader: string | undefined,
  campaignId: string,
  invitationId: string,
): ApiResult {
  const actor = authenticate(authHeader);
  if (!isActor(actor)) return actor;

  const db = getDb();
  const campaign = findCampaign(db, campaignId);
  if (isApiResult(campaign)) return campaign;

  const row = db
    .prepare(
      'SELECT sequence, invitation_id, username, character_id, status FROM play_campaign_invitations WHERE campaign_id = ? AND invitation_id = ?',
    )
    .get(campaignId, invitationId) as InvitationRow | undefined;
  if (!row) {
    return { status: 404, body: { error: 'invitation not found' } };
  }

  if (actor.username !== row.username) {
    return { status: 403, body: { error: 'only the invited user may accept this invitation' } };
  }

  if (row.status !== 'pending') {
    return { status: 409, body: { error: 'invitation is no longer pending' } };
  }

  db.prepare(
    'UPDATE play_campaign_invitations SET status = ? WHERE campaign_id = ? AND invitation_id = ?',
  ).run('accepted', campaignId, invitationId);

  db.prepare(
    'INSERT INTO play_campaign_members (campaign_id, username, character_id, name, class) VALUES (?, ?, ?, ?, ?)',
  ).run(campaignId, row.username, row.character_id, row.username, 'adventurer');

  return {
    status: 200,
    body: {
      invitation_id: row.invitation_id,
      username: row.username,
      character_id: row.character_id,
      status: 'accepted',
    } as unknown as JsonValue,
  };
}

export function listInvitations(authHeader: string | undefined, campaignId: string): ApiResult {
  const actor = authenticate(authHeader);
  if (!isActor(actor)) return actor;

  const db = getDb();
  const campaign = findCampaign(db, campaignId);
  if (isApiResult(campaign)) return campaign;

  const forbidden = requireParticipant(db, campaign, actor, 'not a campaign member');
  const isDm = campaign.owner === actor.username;

  const rows = db
    .prepare(
      'SELECT sequence, invitation_id, username, character_id, status FROM play_campaign_invitations WHERE campaign_id = ? ORDER BY sequence ASC',
    )
    .all(campaignId) as unknown as InvitationRow[];

  if (isDm) {
    return { status: 200, body: { invitations: rows.map(toInvitationObject) } as unknown as JsonValue };
  }

  const own = rows.filter((row) => row.username === actor.username);
  if (own.length > 0) {
    return { status: 200, body: { invitations: own.map(toInvitationObject) } as unknown as JsonValue };
  }

  if (forbidden) return forbidden;

  return { status: 200, body: { invitations: [] } as unknown as JsonValue };
}
